module/zfs/vdev_removal.c

```diff
@@ -2096,6 +2157,114 @@ spa_vdev_remove_top_check(vdev_t *vd)
 	return (0);
 }
 
+int
+spa_vdev_alloc(spa_t *spa, uint64_t guid)
+{
+	vdev_t *vd;
+	uint64_t txg;
+	int error = 0;
+
+	ASSERT(!MUTEX_HELD(&spa_namespace_lock));
+	ASSERT(spa_writeable(spa));
+
+	/* XXX - is this necessary for activate? */
+	txg = spa_vdev_enter(spa);
+
+	ASSERT(MUTEX_HELD(&spa_namespace_lock));
+
+	vd = spa_lookup_by_guid(spa, guid, B_FALSE);
+
+	if (vd == NULL)
+		error = SET_ERROR(ENOENT);
+	else if (vd->vdev_mg == NULL)
+		error = SET_ERROR(ENOTSUP);
+	else if (!vd->vdev_removing)
+		vdev_activate(vd);
+
+	if (error == 0) {
+		vdev_dirty_leaves(vd, VDD_DTL, txg);
```

Comment:
Do we want to remove the XXX and leave the code for now? Or can we hunt down the cases where we want to call `vdev_dirty_leaves()`. I think one of the questions here is: we call it when we set a device non-allocating (ala removing it), but we are also calling it when we mark it back to allocating. Does the latter make sense to keep at all?

This may be worth investigating. It seems to relate to an interesting thing I ran into last weekend:
A pool with 4 raid-z3 vdevs, and a mirrored slog. One of the slog devices had failed.
I was able to `zpool detach` the failed log device, and everything was fine.
I then `zpool attach`d a new SSD. The remaining original SSD was within the last few % of its life, and likely to fail very soon and cause the pool to perform very poorly as the previous one just had. So I tried to detach the old SSD as well.
However, ZFS wouldn't let me: insufficient replicas. Basically it was saying that removing the old SSD from the mirror could not happen because it had to run a complete resilver to be sure the new SSD had all of the data. For a normal mirror this makes sense, for a log device, maybe less so.

Is this because removing a log device dirties the vdev? Does that make sense for a log?
Waiting for a full resilver of 400TB (basically a scrub since none of the regular vdevs were replaced) to remove a log device was a bit impractical.

Reply:
@allanjude I suspect that there is no real differentiation between slog mirrors and "normal mirrors" in this case. Since there may be log data on the old device (e.g., from an unmounted dataset), it needs to be "resilvered". You could probably speed this up my simply removing the entire slog and then adding back the new device.

WRT `vdev_dirty_leaves()`, I see at lease one path (`spa_vdev_remove_cancel()`) where we do *not* do the call to `vdev_dirty_leaves()`. It seems unnecessary to me on the "activate" path... and I suspect it may also be unnecessary on the "passivate" path as well (although we currently call it in all paths). I say we remove it from the activate paths and see if anything shows up in testing... @ahrens, your thoughts?

Reply:
I agree, it would be best to remove these VDD_DTL flags and test, it should "just work".  We can probably also remove the entire call to vdev_dirty_leaves(), since we don't manipulate the leaf state at all.

Reply:
After discussing it with Mark, we think the best idea is to leave it for now, and open an issue to track testing if we can just remove it entirely.